### PR TITLE
Accessibility Fixes Batch 3 - Several Fixes for PackageDetail part of PM UI

### LINKED ISSUE #9157: Narrator does not announce "Blank","No Dependencies","nuget.org","MIT" link label
## batch 3
- [994456](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/994456) - PackageDetail - Tags - A11y_NuGetClient-MinorUIChanges_TestPackageDeprecationOther_ScreenReader : Narrator does not announces "Blank" for the blank field.
- [994459](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/994459) - PackageDetail - A11y_NuGetClient-MinorUIChanges_TestPackage.Deprecation.Other_Installed_ScreenReader : Narrator is annoucing "Dependencies" for "No Dependencies" text.

## batch 4
- [994452](https://devdiv.visualstudio.com/DevDiv/_queries/edit/994452) - PackageItem - A11y_NuGetClient-MinorUIChanges_Browse_ScreenReader : Screen reader focus is not going on the  "version and Package Description" text.

## batch 5
-[1049215](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1049215) - PackageDetail -nuget.org link- A11y_VS ASP.NET and Web Development_ Install Nuget Packages _ Screen Reader: Narrator is not reading the link text for “nuget.org” link
-[994495](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/994495) - PackageDetail - LicenseUI - A11y_NuGetClient-MinorUIChanges_TestPackage.Deprecation.CriticalBugs.AlternatePackage_Installed_ScreenReader : Narrator is not announcing the label association for the "MIT" link.

### PR DESCRIPTION
**Pull request by rrelyea**
## Bug

- Fixes: https://github.com/nuget/home/issues/9157 - batch 3 of fixes
- Fixes: [994456](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/994456) - PackageDetail - Tags - A11y_NuGetClient-MinorUIChanges_TestPackageDeprecationOther_ScreenReader : Narrator does not announces "Blank" for the blank field.
- Fixes: [994459](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/994459) - PackageDetail - A11y_NuGetClient-MinorUIChanges_TestPackage.Deprecation.Other_Installed_ScreenReader : Narrator is annoucing "Dependencies" for "No Dependencies" text.

* Regression: No  
* How are we preventing it in future:   using Accessibility Insights when changing UI

### For NullToVisibilityConverter.cs:
  To support fixing Tags in PackageMetadataControl.xaml, added ability to support empty strings as well as null strings.

### For DetailedPackageMetadata.cs
  To support showing "No Dependencies" if absolutely no TFMs/dependencies, as part of viewmodel, instead of part of XAML tree.

### For PackageDependencyMetadata.cs
  To support showing "No Dependencies" if dependencies for a TFM, as part of viewmodel, instead of part of XAML tree.

### For PackageDependencySetMetadata.cs
  To support showing "No Dependencies" if dependencies for a TFM, as part of viewmodel, instead of part of XAML tree.

### For Resources.xaml
  To support fixing Tags in PackageMetadataControl.xaml, added ability to support empty strings as well as null strings.

### For: src\NuGet.Clients\NuGet.PackageManagement.UI\Xamls\PackageMetadataControl.xaml...
`(50:10) <CrispImage Name="_licenseWarning" />
` 
Error AIX100: missing autoName, either implied, explicit, or labeledBy
Fixed: added AP.Name

Line 84: removed AP.Name to invalid databinding. Text will be exposed as default AP.N for a TextBlock.

`(87:6) <TextBox Name="_description" Content="{Binding ElementName=Description}" CalcAutoName="{Binding ElementName=Description}" />
` 
Error AIX111: don't set name and labeledby, pick one
  Error AIX222: autoname doesn't need to be set. Value is same as content, so it is a wasteful noop.
Fixed: removed AP.Name

`(129:8) <TextBox Content="{Binding ElementName=Version}" CalcAutoName="{Binding ElementName=Version}" />
`  Error AIX111: don't set name and labeledby, pick one
  Error AIX222: autoname doesn't need to be set. Value is same as content, so it is a wasteful noop.
Fixed: removed AP.Name. LabeledBy is good to keep, pointing to TextBlock.

`(150:8) <TextBox Content="{Binding ElementName=Owners}" CalcAutoName="{Binding ElementName=Owners}" />
` 
Error AIX111: don't set name and labeledby, pick one
  Error AIX222: autoname doesn't need to be set. Value is same as content, so it is a wasteful noop.
Fixed: removed AP.Name. LabeledBy is good to keep, pointing to TextBlock.

`(171:8) <TextBox Content="{Binding ElementName=Authors}" CalcAutoName="{Binding ElementName=Authors}" />
`  
Error AIX111: don't set name and labeledby, pick one
  Error AIX222: autoname doesn't need to be set. Value is same as content, so it is a wasteful noop.
Fixed: removed AP.Name. LabeledBy is good to keep, pointing to TextBlock.

`(193:8) <ItemsControl />
`  
Error AIX100: missing autoName, either implied, explicit, or labeledBy
Nothing done now. Still need to work on bug in this area...Tracked by AzDO bug.

Line 209: removed AP.Name to invalid databinding. Text will be exposed as default AP.N for a TextBlock.

`(211:8) <TextBox Content="{Binding ElementName=DownloadCount}" CalcAutoName="{Binding ElementName=DownloadCount}" />
`  
Error AIX111: don't set name and labeledby, pick one
  Error AIX222: autoname doesn't need to be set. Value is same as content, so it is a wasteful noop.
Fixed: removed AP.Name. LabeledBy is good to keep, pointing to TextBlock.

Line 223: removed AP.Name to invalid databinding. Text will be exposed as default AP.N for a TextBlock.

`(232:8) <TextBox Name="datePublished" Content="{Binding ElementName=Published}" CalcAutoName="{Binding ElementName=Published}" />
`  
Error AIX111: don't set name and labeledby, pick one
  Error AIX222: autoname doesn't need to be set. Value is same as content, so it is a wasteful noop.
Fixed: removed AP.Name. LabeledBy is good to keep, pointing to TextBlock.

Line 278: removed AP.Name to invalid databinding. Text will be exposed as default AP.N for a TextBlock.

Line 298: Tags should work like all other fields...if null (or empty), hide the label and the value.
Line 304: removed AP.Name to invalid databinding. Text will be exposed as default AP.N for a TextBlock.

`(306:8) <TextBox Content="{Binding ElementName=Tags}" CalcAutoName="{Binding ElementName=Tags}" />
`  
Error AIX111: don't set name and labeledby, pick one
  Error AIX222: autoname doesn't need to be set. Value is same as content, so it is a wasteful noop.
Fixed: removed AP.Name. LabeledBy is good to keep, pointing to TextBlock.

`(371:8) <TextBlock Name="_hasDependencies" Content="{Static nuget:Resources.Text_NoDependencies}" CalcAutoName="{Binding ElementName=Text}" />
`  
Error AIX111: don't set name and labeledby, pick one
Fixed: ended up doing major surgery on dependency viewing. replaced a few nested itemscontrols and some clumbsy accessibility patch up with a TreeView -- an accessible control.

### For PackageMetadataControl.xaml.cs
  Added 2 event handlers to track selected treeviewitems, so that when you move focus away from that control, it doesn't keep the selecteditem highlighted.

## Testing/Validation

Tests Added: No
Reason for not adding tests:  Accessibility pattern checking with Accessibility Insights is the plan.
Validation:  Accessibility Insights & narrator validation.

### CODE REVIEW

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailedPackageMetadata.cs [hunk 1]
```diff
@@ -32,11 +32,19 @@ public DetailedPackageMetadata(IPackageSearchMetadata serverData, PackageDepreca
...
             DownloadCount = downloadCount;
             Published = serverData.Published;
-            DependencySets = serverData.DependencySets?
-                .Select(e => new PackageDependencySetMetadata(e))
-                ?? new PackageDependencySetMetadata[] { };
-            HasDependencies = DependencySets.Any(
-                dependencySet => dependencySet.Dependencies != null && dependencySet.Dependencies.Count > 0);
+
+            var dependencySets = serverData.DependencySets;
+            if (dependencySets.Any())
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailedPackageMetadata.cs]: The old code guarded against the possibility that `serverData.DependencySets` might be `null`:
```C#
DependencySets = serverData.DependencySets?
```
This new code does not.  `Any()` will happily attempt to dereference `null` if you permit it.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailedPackageMetadata.cs [hunk 2]
```diff
@@ -32,11 +32,19 @@ public DetailedPackageMetadata(IPackageSearchMetadata serverData, PackageDepreca
             Tags = serverData.Tags;
             DownloadCount = downloadCount;
             Published = serverData.Published;
-            DependencySets = serverData.DependencySets?
-                .Select(e => new PackageDependencySetMetadata(e))
-                ?? new PackageDependencySetMetadata[] { };
-            HasDependencies = DependencySets.Any(
-                dependencySet => dependencySet.Dependencies != null && dependencySet.Dependencies.Count > 0);
+
+            var dependencySets = serverData.DependencySets;
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailedPackageMetadata.cs]: Use the explicit type instead of `var`.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailedPackageMetadata.cs [hunk 3]
```diff
@@ -32,11 +32,19 @@ public DetailedPackageMetadata(IPackageSearchMetadata serverData, PackageDepreca
...
+            if (dependencySets.Any())
+            {
+                DependencySets = dependencySets.OrderBy((ds) => ds.TargetFramework.DotNetFrameworkName)
+                    .OrderByDescending((ds) => ds.TargetFramework.Version)
+                    .Select(e => new PackageDependencySetMetadata(e));
+            }
+            else
+            {
+                DependencySets = new PackageDependencySetMetadata[] { new PackageDependencySetMetadata(null) }; //placeholder "No Dependencies"
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailedPackageMetadata.cs]: How about declaring:
```C#
private static readonly IReadOnlyList<PackageDependencySetMetadata> NoDependenciesPlaceholder = new PackageDependencySetMetadata[] { new PackageDependencySetMetadata(dependencyGroup: null) };
```
Then, reference it here:
```C#
DependencySets = NoDependenciesPlaceholder;
```
Advantages:
1.  Fewer allocations.
2.  The named parameter explains the mysterious `null` argument.
3.  The comment becomes unnecessary and can be removed.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageDependencyMetadata.cs [hunk 4]
```diff
@@ -22,6 +22,8 @@ public PackageDependencyMetadata(Packaging.Core.PackageDependency serverData)
 
         public VersionRange Range { get; }
 
+        public bool IsNoDependencyPlaceHolder {get; set;}
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageDependencyMetadata.cs]: Format document.

Make setter `internal`.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageDependencySetMetadata.cs [hunk 5]
```diff
@@ -12,14 +12,47 @@ public class PackageDependencySetMetadata
...
+                    var dependenciesList = new List<PackageDependencyMetadata>();
+                    dependenciesList.Add(dependencyPlaceHolder);
+                    Dependencies = dependenciesList.AsReadOnly();
+                }
+            }
         }
 
         public NuGetFramework TargetFramework { get; private set; }
         public IReadOnlyCollection<PackageDependencyMetadata> Dependencies { get; private set; }
+        public bool IsNoDependencyPlaceHolder { get; set; }
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageDependencySetMetadata.cs]: It looks like the setter can be `private`.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageDependencySetMetadata.cs [hunk 6]
```diff
@@ -12,14 +12,47 @@ public class PackageDependencySetMetadata
...
+                        .ToList()
+                        .AsReadOnly();
+                }
+                else
+                {
+                    // There are no dependencies, instead return a collection with a special place holder, so UI can easily display "No Dependencies"
+                    var dependencyPlaceHolder = new PackageDependencyMetadata() { IsNoDependencyPlaceHolder = true };
+                    var dependenciesList = new List<PackageDependencyMetadata>();
+                    dependenciesList.Add(dependencyPlaceHolder);
+                    Dependencies = dependenciesList.AsReadOnly();
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageDependencySetMetadata.cs]: This allocates a `List<PackageDependencyMetadata>` and then a `ReadOnlyCollection<PackageDependencyMetadata>` wrapping the list.  This happens for each package hitting this code path.

All you need is an `IReadOnlyList<PackageDependencyMetadata>`.  You can achieve this with fewer allocations with a simple array:
```C#
private static readonly IReadOnlyCollection<PackageDependencyMetadata> NoDependenciesPlaceholder = new PackageDependencyMetadata[] { new PackageDependencyMetadata() { IsNoDependencyPlaceHolder = true } };
```
Then, here:
```C#
Dependencies = NoDependenciesPlaceholder;
```
Then you can delete the comment.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageMetadataControl.xaml [hunk 7]
```diff
@@ -315,67 +304,38 @@ Background="{DynamicResource {x:Static nuget:Brushes.DetailPaneBackground}}"
...
+        x:Name="_dependencies"
+        Grid.Row="3"
+        Margin="-21,0,0,0"
+        BorderThickness="0"
+        Background="{x:Null}"
+        TreeViewItem.Selected="OnItemSelected"
+        LostFocus="TreeView_LostFocus"
+        AutomationProperties.Name="{x:Static nuget:Resources.Label_Dependencies}">
+      <TreeView.Template>
+        <!-- remove srollviewer so that mousewheel over treeview content work while over treeview -->
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageMetadataControl.xaml]: scrollviewer

Also, clarify what you're doing.  You're _replacing_ not removing, right?

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageDependencySetMetadata.cs [hunk 8]
```diff
@@ -12,14 +12,47 @@ public class PackageDependencySetMetadata
...
+                }
+            }
         }
 
         public NuGetFramework TargetFramework { get; private set; }
         public IReadOnlyCollection<PackageDependencyMetadata> Dependencies { get; private set; }
+        public bool IsNoDependencyPlaceHolder { get; set; }
+        public string TargetFrameworkDisplay
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageDependencySetMetadata.cs]: I would much rather see this as:
```C#
public string TargetFrameworkDisplay { get; private set; }
```
Then [here](https://github.com/NuGet/NuGet.Client/pull/3229/files#diff-f6fe35ae56cabf8941953d92b67006c5R16-R18) set:
```C#
TargetFrameworkDisplay = Resources.Text_NoDependencies;
```
...and [here](https://github.com/NuGet/NuGet.Client/pull/3229/files#diff-f6fe35ae56cabf8941953d92b67006c5R20-R36) set:
```C#
TargetFrameworkDisplay = TargetFramework.ToString();
```
This will avoid the possibility of `TargetFramework.ToString()` being called multiple times.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageDependencyMetadata.cs [hunk 9]
```diff
@@ -30,7 +32,11 @@ public PackageDependencyMetadata(string id, VersionRange range)
 
         public override string ToString()
         {
-            if (Range == null)
+            if (IsNoDependencyPlaceHolder)
+            {
+                return Resources.Text_NoDependencies;
+            }
+            else if (Range == null)
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageDependencyMetadata.cs]: You don't need `else`.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/NullToVisibilityConverter.cs [hunk 10]
```diff
@@ -26,5 +33,10 @@ public object ConvertBack(object value, Type targetType, object parameter, Cultu
             Debug.Fail("Not Implemented");
             return null;
         }
+
+        /// <summary>
+        /// If a string is null or empty, collapse it. By default, it will only collapse null values (including for strings)
+        /// </summary>
+        public bool CollapseEmptyString { get; set; }
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/NullToVisibilityConverter.cs]: Can the setter be `internal`?

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/NullToVisibilityConverter.cs [hunk 11]
```diff
@@ -15,7 +15,14 @@ public object Convert(object value, Type targetType, object parameter, CultureIn
         {
             if (targetType == typeof(Visibility))
             {
-                return value == null ? Visibility.Collapsed : Visibility.Visible;
+                if (CollapseEmptyString && value.GetType() == typeof(string))
+                {
+                    return string.IsNullOrEmpty((string)value) ? Visibility.Collapsed : Visibility.Visible;
+                }
+                else
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/NullToVisibilityConverter.cs]: Like the parent `if` statement, an `else` block is unnecessary.
```C#
if (…)
{
    return …;
}

return …;
```

**Review by dtivel — CHANGES_REQUESTED** [review]

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/NullToVisibilityConverter.cs [hunk 12]
```diff
@@ -15,7 +15,14 @@ public object Convert(object value, Type targetType, object parameter, CultureIn
         {
             if (targetType == typeof(Visibility))
             {
-                return value == null ? Visibility.Collapsed : Visibility.Visible;
+                if (CollapseEmptyString && value.GetType() == typeof(string))
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/NullToVisibilityConverter.cs]: Something seems wrong.

This class is `NullToVisibilityConverter` which strongly implies that it converts `null` values to some `Visibility` value and that `value` will be `null`.

If so, `value.GetType()` will generate a `NullReferenceException`.  Sure, you could first check if `value` is `null` like the original code did, but then this class seems to be the wrong place for converting a non-`null` value to a `Visibility` value.

What's wrong?

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/NullToVisibilityConverter.cs (hunk 10 above)
Comment (rrelyea) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/NullToVisibilityConverter.cs]: to be accessed by xaml, members generally need to be public. Converters are generally built to be used. 
I'm refreshing my memory to see if it is technically possible...it might be: https://docs.microsoft.com/en-us/dotnet/desktop-wpf/xaml-services/define-custom-types#access-level

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailedPackageMetadata.cs (hunk 3 above)
Comment (rrelyea) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailedPackageMetadata.cs]: love it.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageDependencyMetadata.cs (hunk 9 above)
Comment (rrelyea) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageDependencyMetadata.cs]: yes, but it makes me feel good. doesn't c# compiler optimize it away?

Comment (rrelyea) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageDependencyMetadata.cs]: fixing.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageDependencySetMetadata.cs (hunk 8 above)
Comment (rrelyea) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageDependencySetMetadata.cs]: generally, wpf databinding will only call that property again, if the object implements INotifyPropertyChange and if the property does change.  Let me think through the advice given that and after i view what else this class is used from.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageMetadataControl.xaml (hunk 7 above)
Comment (rrelyea) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageMetadataControl.xaml]: Replacing with:
          Replaces default control template for this treeview with a slimmer control template.
          Specifically, this one has no ScrollViewer...which wasn't being used, and was making mousewheel not work for
          the larget detail scrollable area while the mouse was over the treeview.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/NullToVisibilityConverter.cs (hunk 10 above)
Comment (rrelyea) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/NullToVisibilityConverter.cs]: i tested just switching it...and it does work...(me in a former life helped make that possible).
But since you dropped by and said, not important, i reverted it.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageDependencySetMetadata.cs (hunk 8 above)
Comment (rrelyea) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageDependencySetMetadata.cs]: ended up going with your suggestions. i like getting rid of that bool too.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/NullToVisibilityConverter.cs (hunk 12 above)
Comment (rrelyea) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/NullToVisibilityConverter.cs]: yes, good point...my new code path doesn't work for null values. It only works for empty strings. 
it should support both.
In our usage, it appears that Tags is always empty, not null...so i didn't catch that with runtime testing.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/NullToVisibilityConverter.cs [hunk 13]
```diff
@@ -15,7 +15,17 @@ public object Convert(object value, Type targetType, object parameter, CultureIn
...
             {
-                return value == null ? Visibility.Collapsed : Visibility.Visible;
+                if (value == null)
+                {
+                    return Visibility.Collapsed;
+                }
+
+                if (CollapseEmptyString && value.GetType() == typeof(string))
+                {
+                    return (string)value == string.Empty ? Visibility.Collapsed : Visibility.Visible;
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/NullToVisibilityConverter.cs]:
```C#
return string.IsNullOrEmpty((string)value) ? Visibility.Collapsed : Visibility.Visible;
```
See [this](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1820?view=vs-2019).

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailedPackageMetadata.cs [hunk 14]
```diff
@@ -32,11 +32,19 @@ public DetailedPackageMetadata(IPackageSearchMetadata serverData, PackageDepreca
...
-                .Select(e => new PackageDependencySetMetadata(e))
-                ?? new PackageDependencySetMetadata[] { };
-            HasDependencies = DependencySets.Any(
-                dependencySet => dependencySet.Dependencies != null && dependencySet.Dependencies.Count > 0);
+
+            IEnumerable<PackageDependencyGroup> dependencySets = serverData.DependencySets;
+            if (dependencySets != null && dependencySets.Any())
+            {
+                DependencySets = dependencySets.OrderBy((ds) => ds.TargetFramework.DotNetFrameworkName)
+                    .OrderByDescending((ds) => ds.TargetFramework.Version)
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailedPackageMetadata.cs]: The `OrderBy(…).OrderByDescending(…)` seems unrelated to accessibility fixes.  It seems like it should be a separate issue/PR.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailedPackageMetadata.cs [hunk 15]
```diff
@@ -32,11 +32,19 @@ public DetailedPackageMetadata(IPackageSearchMetadata serverData, PackageDepreca
...
+                    .Select(e => new PackageDependencySetMetadata(e));
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailedPackageMetadata.cs]: Can `DependencySets` be read multiple times?  If yes, I think you should add `.ToArray()` for reasons similar to this [statement](https://github.com/NuGet/NuGet.Client/pull/3229/files#diff-f6fe35ae56cabf8941953d92b67006c5R26-R29).

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/NullToVisibilityConverter.cs (hunk 13 above)
Comment (rrelyea) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/NullToVisibilityConverter.cs]: I'm adding a change to DetailedPackageMetadata.cs Tags property to be null if null or empty.
Then we can get rid of all these changes to the NullVisibilityConverter.cs.

More elegant way to fix a data layer problem...in the data layer.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailedPackageMetadata.cs (hunk 15 above)
Comment (rrelyea) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailedPackageMetadata.cs]: next delta will have this.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailedPackageMetadata.cs (hunk 14 above)
Comment (rrelyea) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailedPackageMetadata.cs]: Will remove this for now... ".OrderBy((ds) => ds.TargetFramework.DotNetFrameworkName)
                    .OrderByDescending((ds) => ds.TargetFramework.Version)" and merge into dev seperately.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/NullToVisibilityConverter.cs [hunk 16]
```diff
@@ -1,4 +1,4 @@
-﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Copyright (c) .NET Foundation. All rights reserved.
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/NullToVisibilityConverter.cs]: Remove this file from your PR.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailedPackageMetadata.cs [hunk 17]
```diff
@@ -29,14 +29,22 @@ public DetailedPackageMetadata(IPackageSearchMetadata serverData, PackageDepreca
             LicenseUrl = serverData.LicenseUrl;
             ProjectUrl = serverData.ProjectUrl;
             ReportAbuseUrl = serverData.ReportAbuseUrl;
-            Tags = serverData.Tags;
+            // Some server implementations send down an array with an empty string, which ends up as an empty string.
+            // In PM UI, we want Tags to work like most other properties from the server (Authors/Owners), and be null, if there is no value.
+            Tags = string.IsNullOrEmpty(serverData.Tags) ? null : serverData.Tags ;
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailedPackageMetadata.cs]: Format document.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageMetadataControl.xaml [hunk 18]
```diff
@@ -315,67 +304,42 @@ Background="{DynamicResource {x:Static nuget:Brushes.DetailPaneBackground}}"
...
+          <Setter Property="IsExpanded" Value="true" />
+          <Setter Property="AutomationProperties.Name" Value="" />
+          <Setter Property="Foreground" Value="{DynamicResource {x:Static nuget:Brushes.UIText}}" />
+          <Setter Property="FontWeight" Value="normal" />
+        </Style>
+        <HierarchicalDataTemplate ItemsSource="{Binding Dependencies}" DataType="{x:Type nuget:PackageDependencySetMetadata}">
+          <TextBlock Text="{Binding TargetFrameworkDisplay}" Padding="0,2,0,0" />
+        </HierarchicalDataTemplate>
+        <DataTemplate DataType="{x:Type nuget:PackageDependencyMetadata}">
+          <TextBlock Text="{Binding }" ToolTip="{Binding }" />
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageMetadataControl.xaml]: Is the extra space significant?  `{Binding }` vs. `{Binding}`?

**Review by dtivel — APPROVED** [review]

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageMetadataControl.xaml (hunk 18 above)
Comment (rrelyea) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageMetadataControl.xaml]: nope identical. markup extensions (like binding), can have params. but this doesn't. likely cleaner to remove space.

**Review by dtivel — APPROVED** [review]

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageMetadataControl.xaml [hunk 19]
```diff
@@ -315,67 +304,41 @@ Background="{DynamicResource {x:Static nuget:Brushes.DetailPaneBackground}}"
...
+          the larget detail scrollable area while the mouse was over the treeview.
+        -->
+        <ControlTemplate TargetType="TreeView">
+          <ItemsPresenter />
+        </ControlTemplate>
+      </TreeView.Template>
+      <TreeView.Resources>
+        <Style TargetType="TreeViewItem">
+          <Setter Property="IsExpanded" Value="true" />
+          <Setter Property="AutomationProperties.Name" Value="" />
```
Comment (dominoFire) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageMetadataControl.xaml]: Is this intentional ?

Comment (dominoFire) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageMetadataControl.xaml]: When you read with the narrator in Caps Lock + arrow, it reads each tree view item as the type of the tree view item: "NuGet.PackageManagement.UI.PackageDependencySetMetadata..."

image

Possible fixes:

- Remove the line
- Provide a more friendly name, for example, `"Target Framework {tfm}, version {version}"`

Comment (rrelyea) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageMetadataControl.xaml]: Will dig into this behavior and options. Looks to me like it is exposing some details about the data object. the other data is already exposed naturally in the content under it (header and children), so I'd prefer not to repeat.

Comment (rrelyea) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageMetadataControl.xaml]: fixed with workaround...and opened a WPF bug: https://github.com/dotnet/wpf/issues/2552

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageMetadataControl.xaml.cs [hunk 20]
```diff
@@ -69,5 +69,24 @@ private void PackageMetadataControl_DataContextChanged(object sender, Dependency
...
+            _dependencies.Tag = e.OriginalSource;
+        }
+
+        private void TreeView_LostFocus(object sender, RoutedEventArgs e)
+        {
+            // hide focus highlight when treeview lostfocus
+            if (_dependencies.SelectedItem != null)
+            {
+                TreeViewItem selectedTVI = _dependencies.Tag as TreeViewItem;
+                if (selectedTVI != null)
```
Comment (dominoFire) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageMetadataControl.xaml.cs:85]: Casting and checking in one line:

```suggestion
                if (_dependencies.Tag is TreeViewItem selectedTVI)
```

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageMetadataControl.xaml [hunk 21]
```diff
@@ -315,67 +304,49 @@ Background="{DynamicResource {x:Static nuget:Brushes.DetailPaneBackground}}"
...
+        </ControlTemplate>
+      </TreeView.Template>
+      <TreeView.Resources>
+        <nuget:AccessibleConverter x:Key="AccNameWorkaround" />
+        <HierarchicalDataTemplate ItemsSource="{Binding Dependencies}" DataType="{x:Type nuget:PackageDependencySetMetadata}">
+          <TextBlock Text="{Binding TargetFrameworkDisplay}" Padding="0,2,0,0" />
+        </HierarchicalDataTemplate>
+        <DataTemplate DataType="{x:Type nuget:PackageDependencyMetadata}">
+          <TextBlock Text="{Binding}" ToolTip="{Binding}" />
+        </DataTemplate >
```
Comment (dominoFire) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageMetadataControl.xaml]:
```suggestion
        </DataTemplate>
```

**Review by dominoFire — APPROVED** [review]
LGTM. I left some style and typo comments.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageMetadataControl.xaml.cs (hunk 20 above)
Comment (rrelyea) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageMetadataControl.xaml.cs:85]: Will do with general clean up post Dev->Release branch. Thx.
